external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaRDD.scala

```diff
@@ -0,0 +1,220 @@
+    " This should not happen, and indicates that messages may have been lost"
+
+  private def errOvershotEnd(itemOffset: Long, part: KafkaRDDPartition): String =
+    s"Got ${itemOffset} > ending offset ${part.untilOffset} " +
+    s"for topic ${part.topic} partition ${part.partition} start ${part.fromOffset}." +
+    " This should not happen, and indicates a message may have been skipped"
+
+  override def compute(thePart: Partition, context: TaskContext): Iterator[R] = {
+    val part = thePart.asInstanceOf[KafkaRDDPartition]
+    assert(part.fromOffset <= part.untilOffset, errBeginAfterEnd(part))
+    if (part.fromOffset == part.untilOffset) {
+      log.warn("Beginning offset ${part.fromOffset} is the same as ending offset " +
+        s"skipping ${part.topic} ${part.partition}")
+      Iterator.empty
+    } else {
+      new KafkaRDDIterator(part, context)
+    }
+  }
+
+  private class KafkaRDDIterator(
+      part: KafkaRDDPartition,
+      context: TaskContext) extends NextIterator[R] {
+
+    context.addTaskCompletionListener{ context => closeIfNeeded() }
+
+    log.info(s"Computing topic ${part.topic}, partition ${part.partition} " +
+      s"offsets ${part.fromOffset} -> ${part.untilOffset}")
+
+    val kc = new KafkaCluster(kafkaParams)
+    val keyDecoder = classTag[U].runtimeClass.getConstructor(classOf[VerifiableProperties])
+      .newInstance(kc.config.props)
+      .asInstanceOf[Decoder[K]]
+    val valueDecoder = classTag[T].runtimeClass.getConstructor(classOf[VerifiableProperties])
+      .newInstance(kc.config.props)
+      .asInstanceOf[Decoder[V]]
+    val consumer = connectLeader
+    var requestOffset = part.fromOffset
+    var iter: Iterator[MessageAndOffset] = null
+
+    // The idea is to use the provided preferred host, except on task retry atttempts,
+    // to minimize number of kafka metadata requests
+    private def connectLeader: SimpleConsumer = {
+      if (context.attemptNumber > 0) {
+        kc.connectLeader(part.topic, part.partition).fold(
+          errs => throw new SparkException(
+            s"Couldn't connect to leader for topic ${part.topic} ${part.partition}: " +
+              errs.mkString("\n")),
+          consumer => consumer
+        )
+      } else {
+        kc.connect(part.host, part.port)
+      }
+    }
+
+    private def handleFetchErr(resp: FetchResponse) {
+      if (resp.hasError) {
+        val err = resp.errorCode(part.topic, part.partition)
+        if (err == ErrorMapping.LeaderNotAvailableCode ||
+          err == ErrorMapping.NotLeaderForPartitionCode) {
+          log.error(s"Lost leader for topic ${part.topic} partition ${part.partition}, " +
+            s" sleeping for ${kc.config.refreshLeaderBackoffMs}ms")
+          Thread.sleep(kc.config.refreshLeaderBackoffMs)
+        }
+        // Let normal rdd retry sort out reconnect attempts
+        throw ErrorMapping.exceptionFor(err)
+      }
+    }
+
```

Comment:
My documentation for the stream already says you need to have sufficient
Kafka log retention.  The Kafka documentation for OffsetOutOfRangeException
already says it Indicates the client has requested a range no longer
available on the server.

The stream stopping is exactly what should happen in that case.  Then its
the clients choice as to how to handle the situation.  If someone really
wants to handle the fact that they underprovisioned Kafka by just
restarting the stream at the head of the log and losing data, they can do
so by starting up the stream again with auto.offset.reset set to largest
(which is also already documented). But they should have to do that
themselves, we shouldn't be building a library that silently throws away
data in failure situations.

I think that's fairly clear, but if you want to write up something for
including in the streaming programming guide I'll take a look.
On Jan 31, 2015 11:27 AM, "Dibyendu Bhattacharya" <notifications@github.com>
wrote:

> In
> external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaRDD.scala
> <https://github.com/apache/spark/pull/3798#discussion_r23890423>:
>
> > +    }
> > +
> > +    private def handleFetchErr(resp: FetchResponse) {
> > +      if (resp.hasError) {
> > +        val err = resp.errorCode(part.topic, part.partition)
> > +        if (err == ErrorMapping.LeaderNotAvailableCode ||
> > +          err == ErrorMapping.NotLeaderForPartitionCode) {
> > +          log.error(s"Lost leader for topic ${part.topic} partition ${part.partition}, " +
> > +            s" sleeping for ${kc.config.refreshLeaderBackoffMs}ms")
> > +          Thread.sleep(kc.config.refreshLeaderBackoffMs)
> > +        }
> > +        // Let normal rdd retry sort out reconnect attempts
> > +        throw ErrorMapping.exceptionFor(err)
> > +      }
> > +    }
> > +
>
> Hi Cody, I believe in this implementation the range of offset of given
> only at the start, for rest of the flow the offset range is calculated
> automatically. What I say, if Offset_Out_Of_Range comes in those RDDs where
> you calculate the offset, this implementation can not recover from this
> failure. I did not say that we always get this error all time , but this
> error is not very rare cases also. and if the Receiver stops because of
> this, that's an issue.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/3798/files#r23890423>.
>